hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/TokenPreTransactionHandlerImpl.java

```diff
@@ -172,14 +173,47 @@ public TransactionMetadata preHandleUpdateTokenFeeSchedule(
     public TransactionMetadata preHandlePauseToken(
             @NonNull final TransactionBody txn, @NonNull final AccountID payer) {
         Objects.requireNonNull(txn);
-        throw new NotImplementedException();
+        final var op = txn.getTokenPause();
+        final var meta =
+                new SigTransactionMetadataBuilder(accountStore).payerKeyFor(payer).txnBody(txn);
+
+        if (op.hasToken()) {
+            addPauseKey(op.getToken(), meta);
+        } else {
+            meta.status(ResponseCodeEnum.INVALID_TOKEN_ID);
+        }
+
+        return meta.build();
     }
 
     @Override
     /** {@inheritDoc} */
     public TransactionMetadata preHandleUnpauseToken(
             @NonNull final TransactionBody txn, @NonNull final AccountID payer) {
         Objects.requireNonNull(txn);
-        throw new NotImplementedException();
+        final var op = txn.getTokenUnpause();
+        final var meta =
+                new SigTransactionMetadataBuilder(accountStore).payerKeyFor(payer).txnBody(txn);
+
+        if (op.hasToken()) {
+            addPauseKey(op.getToken(), meta);
```

Comment:
It seems that calling `addPauseKey` for an "unpause" handler is weird. Is this expected?

Reply:
yeah, token's pause key is used to sign for `tokenPause` and `tokenUnpause`.